### ISSUE TITLE
Local Sharing That Works Like LastPass (for Team with Network Shares)

### ISSUE THREAD

**Issue by MountainX** [keepassxreboot/keepassxc#4091]
# Summary

We have a team of 11 people. Some are managers with access to important account information, others are non-privileged users who do not need access to company financial records, etc.

We would like to be able to use KeePassXC for sharing passwords internally. We want to move away from cloud-based password managers (such as LastPass).

In LastPass you can share any password with one or more people on your team and that password is only available to the specific people you share it with.

Currently, as I understand it, all passwords in any shared KeePassXC database are available to anyone who has the main password to that database. When a team has access to that shared database it means anything that is shared is shared with all. Therefore, **one shared database does not meet our requirements.** 

Could we accomplish our goal with KeePassXC and multiple shared databases? It appears not, as I will explain in the next section.

## The Problem
Sometimes I want to share a bank login with two of my managers, but not with anyone else. Or sometimes I want to share a certain website login with only four team members. There are some logins I want to share with everyone, but those are rare.

In LastPass, when I share an item I can pick from a list of people on my team and easily share the item with exactly the intended individuals.

The current implementation of KeePass gives anyone with the shared database password full access to every password in that database. Therefore, for John and Jane to share passwords only with each other, they need a dedicated database. For John and Raj to share passwords only with each other, they also need a dedicated database. For Liz to share passwords with the managers, another dedicated database is required. For me to share passwords only with managers, yet another database is required. 

If I understand things correctly, for each one-to-one sharing relationship a separate database is required. The one-to-one relationship could be between two people or two groups or a person and a group. But in each case, I believe the current KeeShare implementation requires a separate database for each of those sharing contexts.

For a team of 11 people with 6 roles (managers, financial, sales, engineering, shipping plus an everyone group) a total of **136 different databases would be required**: 17!/(2!(17-2)!). I think 136 shared databases is impossible to maintain for a small team of 11 people. And when a 12th person joins, I wouldn't even know how to integrate them.

## Desired Behavior

I would like to create and configure **a single shared database** (which resides on a local network share). Instead of my shared passwords traveling through LastPass's cloud, they would be shared through the database internal to our network.

In that database I would like to be able to add all my team members as users and define groups such as managers, financial, sales, engineering, shipping plus an everyone group. 

I would like to give everyone on my team access to that one shared database on the network share and let them open it in KeePassXC (along with having their own personal database). 

Within this database, each person would have access only to the passwords shared directly to them

Then each person should be able to share any login credentials with any other member of the team (even outside their role).

Importantly, if I share a login with Jane, only Jane should have access to it. Each shared item should only be available to the individual or group it was shared with (and the person who shared it).

## Possible Solution

I don't know the implementation details. Most password managers allow this type of sharing. Most database systems support user and group permissions. 

It would seem like inside the shared database, you would attach the information about the intended recipient(s) of the shared item, and only give access to those recipients.

## Context

This feature is important for sharing passwords among teams or families (or even roommates, friends, etc.). The situation where you want to always share every password with everyone else on the team is rare. Most likely you want to share a certain password with a specific individual, then you need to share a different password with a different person. Or you want to share a password with a certain group (such as managers). Yes, sometimes you want to share a password with everyone too, but in my experience that is the more rare case.

**Comment by varjolintu** [comment]
Have you considered if KeeShare is enough to cover all the features you need?

There's some documentation available here:
https://github.com/keepassxreboot/keepassxc/blob/develop/docs/QUICKSTART.md#using-sharing

**Comment by droidmonkey** [comment]
I would love to support an enterprise version of KeePassXC that meets the requirements you listed out (and more). Unfortunately at this time that would be a significant departure from the KDBX "standard" set by KeePass and is very complicated without the server-client relationship that LastPass has.

With that said, I do recommend that you think strongly about your scenario and how modern security has evolved. Passing around credentials in a complex web like you describe is a very terrible way of doing business. At the size of your company you should really be looking into try identity and access management with single sign-on, domain credentials, etc. You must be aware that any breach of your sharing will result in reset of credentials with cascading effects (need to redistribute those credentials, change them in the services, etc etc).

**Comment by MountainX** [comment]
@droidmonkey If / when you decide to create an enterprise version of KeePassXC, please let me know.

In terms of security model, I am not convinced that SSO is ready yet. LastPass Enterprise offers single sign-on with over 1200 integrations. I think they may be the leader. It sounds like a lot, but they don't include most of the vendors we do business with. (And I hope to move away from the cloud anyway.)

We do currently use LastPass, and given the limitations of many vendors we deal with, the reality is that we have to share some passwords. The sharing mechanism LastPass offers does meet our requirements except for being in the cloud. I think there is a need for sharing passwords that will continue for some more years.

As far as how to eliminate the need to share passwords, I suppose that depends on other companies. and how fast they implement new technology. Currently, many of our vendors do not even support delegated users. You get one login and only one login for your account. Furthermore, even when I can create delegated user accounts for team members, I often want to set up the account (and ensure a very strong password and other things) then share that login with a team member in a secure manner. And I want to ensure the team member makes proper use of the password manager. So I see a continuing need for password sharing for several more years (maybe longer).

I suppose self-hosted Bitwarden Teams or Enterprise might be an option. But I was hoping there was a way to accomplish our goals with KeePassXC.

If I am way off base on the above, please explain. Thank you for your work on KeePassXC and for such great responses here.

**Comment by droidmonkey** [comment]
Ah understood, I was thinking more internal applications then external to your organization. Without a centralized server it would be impossible to securely regulate which credentials people can see. Even if we encrypted each entry individually, you would have to share a decryption key (password, etc) for each entry to the intended recipient. This is the nature of "offline" password managers. In the cloud scenario, the central service controls the KDBX database and doles out credentials to authorized parties. For us to support this scenario we would have to essentially make a KeePassXC-Server that performed that function and the client applications would interact with that server to retrieve credentials similar to how KeeShare works.

**Comment by MountainX** [comment]
I just learned about Pass: The Standard Unix Password Manager
https://www.passwordstore.org/
It has an implementation for sharing passwords that meets the requirements I described above. I will probably use Pass for sharing passwords for now, but it would be great if KeePassXC decides to incorporate this functionality. My team and I would like to be able to use KeePassXC for everything.

**Comment by MountainX** [comment]
I just learned about Pass: The Standard Unix Password Manager
https://www.passwordstore.org/
It has an implementation for sharing passwords that meets the requirements I described above. The implementation looks to be very simple. It does not depend upon a centralized server. 

> For us to support this scenario we would have to essentially make a KeePassXC-Server that performed that function and the client applications would interact with that server to retrieve credentials similar to how KeeShare works.

Take a look at how Pass does it. It might be simple enough for KeePassXC to add.

I will probably use Pass for sharing passwords for now, but it would be great if KeePassXC decides to incorporate this functionality. My team and I would like to be able to use KeePassXC for everything.
